src/hypothesis/searchstrategy/types.py

```diff
@@ -0,0 +1,260 @@
+    """Return a set of required args in this argspec."""
+    return set(spec.args + spec.kwonlyargs) \
+        - set(spec.args[len(spec.args) - len(spec.defaults or ()):]) \
+        - set(spec.kwonlydefaults or ()) - set(args) - set(kwargs)
+
+
+@st.cacheable
+def type_strategy_mapping():
+    """Return a dict mapping from types to corresponding search strategies.
+
+    Most resolutions will terminate here or in the special handling for
+    generics from the typing module.
+
+    """
+    import uuid
+    import decimal
+    import datetime as dt
+    import fractions
+
+    known_type_strats = {
+        # Types with core Hypothesis strategies
+        type(None): st.none(),
+        bool: st.booleans(),
+        float: st.floats(),
+        complex: st.complex_numbers(),
+        fractions.Fraction: st.fractions(),
+        decimal.Decimal: st.decimals(),
+        text_type: st.text(),
+        bytes: st.binary(),
+        dt.datetime: st.datetimes(),
+        dt.date: st.dates(),
+        dt.time: st.times(),
+        dt.timedelta: st.timedeltas(),
+        uuid.UUID: st.uuids(),
+        # Built-in types
+        type: st.sampled_from([type(None), bool, int, str, list, set, dict]),
+        type(Ellipsis): st.just(Ellipsis),
+        type(NotImplemented): st.just(NotImplemented),
+        bytearray: st.binary().map(bytearray),
+        memoryview: st.binary().map(memoryview),
+        # Pull requests with more types welcome!
+    }
+    for t in integer_types:
+        known_type_strats[t] = st.integers()
+    # build empty collections, as only generics know their contents
+    known_type_strats.update({
+        t: st.builds(t) for t in (tuple, list, set, frozenset, dict)
+    })
+    try:
+        from hypothesis.extra.pytz import timezones
+        known_type_strats[dt.tzinfo] = timezones()
+    except ImportError:  # pragma: no cover
+        pass
+    try:  # pragma: no cover
+        import numpy as np
+        from hypothesis.extra.numpy import \
+            arrays, array_shapes, scalar_dtypes, nested_dtypes
+        known_type_strats.update({
+            np.dtype: nested_dtypes(),
+            np.ndarray: arrays(scalar_dtypes(), array_shapes(max_dims=2)),
+        })
+    except ImportError:
+        pass
+    return known_type_strats
+
+
+def type_sorting_key(t):
+    """Minimise to None, then non-container types, then container types."""
+    if t is None or t is type(None):
+        return -1
+    return issubclass(t, collections.abc.Container)
+
+
+def check_by_origin(thing, super_):
+    return thing is super_ or getattr(thing, '__origin__', None) is super_
+
+
+def try_issubclass(thing, maybe_superclass):
+    try:
+        return issubclass(thing, maybe_superclass)
+    except (AttributeError, TypeError):
+        # Some types can't be the subject or object of an instance or
+        # subclass check, including _TypeAlias and Union
+        return False
+
+
+def from_typing_type(thing):
```

Comment:
So you know that thing I do where I have to write giant blocks of comments in some functions with tricky logic on the grounds that if I don't nobody else will have a clue what the function is doing?

Congratulations! I think you're the first person who isn't me who needs to do this in Hypothesis. 😁 

Because I'm afraid I do not get the logic of this function *at all*.